gitbutler-ui/src/lib/backend/aiService.test.ts

```diff
@@ -0,0 +1,243 @@
+	AnthropicModelName,
+	ConfigKeys,
+	KeyOption,
+	ModelKind,
+	OpenAIModelName
+} from '$lib/backend/aiService';
+import { getCloudApiClient } from '$lib/backend/cloud';
+import * as toasts from '$lib/utils/toasts';
+import { expect, test, describe, vi } from 'vitest';
+import type { AIClient } from '$lib/backend/aiClient';
+import type { GitConfigService } from '$lib/backend/gitConfigService';
+
+const defaultGitConfig = Object.freeze({
+	[ConfigKeys.ModelProvider]: ModelKind.OpenAI,
+	[ConfigKeys.OpenAIKeyOption]: KeyOption.ButlerAPI,
+	[ConfigKeys.OpenAIKey]: undefined,
+	[ConfigKeys.OpenAIModelName]: OpenAIModelName.GPT35Turbo,
+	[ConfigKeys.AnthropicKeyOption]: KeyOption.ButlerAPI,
+	[ConfigKeys.AnthropicKey]: undefined,
+	[ConfigKeys.AnthropicModelName]: AnthropicModelName.Haiku
+});
+
+class DummyGitConfigService implements GitConfigService {
+	constructor(private config: { [index: string]: string | undefined }) {}
+
+	async get<T extends string>(key: string): Promise<T | undefined> {
+		return (this.config[key] || undefined) as T | undefined;
+	}
+
+	async getWithDefault<T extends string>(key: string): Promise<T> {
+		return this.config[key] as T;
+	}
+
+	async set<T extends string>(key: string, value: T): Promise<T | undefined> {
+		return (this.config[key] = value);
+	}
+}
+
+const fetchMock = vi.fn();
+const cloud = getCloudApiClient({ fetch: fetchMock });
+
+class DummyAIClient implements AIClient {
+	constructor(private response = 'lorem ipsum') {}
+
+	async evaluate(_prompt: string) {
+		return this.response;
+	}
+}
+
+const examplePatch = `
+@@ -52,7 +52,8 @@
+
+ export enum AnthropicModelName {
+ 	Opus = 'claude-3-opus-20240229',
+-	Sonnet = 'claude-3-sonnet-20240229'
++	Sonnet = 'claude-3-sonnet-20240229',
++	Haiku = 'claude-3-haiku-20240307'
+ }
+
+ export const AI_SERVICE_CONTEXT = Symbol();
+`;
+
+function buildDefaultAIService() {
+	const gitConfig = new DummyGitConfigService(structuredClone(defaultGitConfig));
+	return new AIService(gitConfig, cloud);
+}
+
+describe.concurrent('AIService', () => {
+	describe.concurrent('#buildModel', () => {
+		test('With default configuration, When a user token is provided. It returns ButlerAIClient', async () => {
+			const aiService = buildDefaultAIService();
+
+			expect(await aiService.buildClient('token')).toBeInstanceOf(ButlerAIClient);
+		});
+
+		test('With default configuration, When a user is undefined. It returns undefined', async () => {
+			const toastErrorSpy = vi.spyOn(toasts, 'error');
+			const aiService = buildDefaultAIService();
+
+			expect(await aiService.buildClient()).toBe(undefined);
+			expect(toastErrorSpy).toHaveBeenLastCalledWith(
+				"When using GitButler's API to summarize code, you must be logged in"
+			);
+		});
+
+		test('When token is bring your own, When a openAI token is present. It returns OpenAIClient', async () => {
+			const gitConfig = new DummyGitConfigService({
+				...defaultGitConfig,
+				'gitbutler.aiOpenAIKeyOption': KeyOption.BringYourOwn,
```

Comment:
Could you still update these strings to point to `ConfigKeys`? On the same note, I didn't necessarily mean to leave it named what it is. Perhaps it would be better to rename it e.g. `GitConfigKey`?

Reply:
Yep, I missed those strings, I'll go and do those.

As for the name, part of me wants to indicate that they're related to the AI functionality and go for `GitAIConfigKey` to differentiate between some other config key enum in the future.

Part of me still likes the idea of having getters and setters for the fields, perhaps now on the GitConfigService, so we can say `await gitConfigService.getOpenAIKeyOption()` (or maybe as a method declared as `get`) so we don't need to worry about being consistent with default values in other areas of the code

Reply:
>await gitConfigService.getOpenAIKeyOption()

How many getters and setters do you think we could eventually end up with?

Reply:
> How many getters and setters do you think we could eventually end up with?

At the minute we've got two wrappers for local storage (persisted.ts & userSettings.ts), the gitbutler.json, various data stored in the new TOML files, and the git config.

I think there are several considerations around reducing the quantity of locations we store data, whether we want to sync the users settings to the could, and deciding on if we want functionality to provide project level overrides like vscode.

In the long run, my thought is that it would be helpful to have one unified interface (even if behind the scenes it writes to different providers) for accessing reading and writing settings so we don't end up with duplicate functionality and inconsistent interfaces.

---

Now is certainly not the time to reinvent that particular wheel. However, by adding these getters and setters, we're going to completely avoid any key name problems, inconsistent default problems, and typing problems in application code. To me it seems like a very sensible thing to do given that the cost of removing this abstraction is going to be pretty much equivalent to finding and replacing the `get` and `set` methods.

Reply:
I've use the enum in the other tests and renamed it.

Perhaps refactoring into setters and getters can be left to another time.

Something I do want to do before we merge in, is create a better metric for when the AI generation button should be disabled, because currently you must be logged in if you're providing your own token, but apart from that, if we're happy from a UI perspective, I'd be happy for this to go ahead and be merged in

Reply:
I've made the change to allow using AI without a user if a key is provided and updated the settings to better reflect that